### ISSUE TITLE
Training goes astray in the second epoch (with --right-left)

### ISSUE THREAD

**Issue by bhaddow** [marian-nmt/marian-dev#240]
Training with approx 60M sentences, using the --right-left flag, I noticed that development scores collapsed just after the start of the second epoch. This happened on 4 identical runs, but did not happen on other (smaller) data sets. I am not using --sqlite

End of first epoch
Ep. 1 : Up. 350000 : Sen. 61128422 : Cost 42.18 : Time 677.69s : 2063.19 words/s

Start of second epoch:
Ep. 2 : Up. 351000 : Sen. 82422 : Cost 84.35 : Time 735.34s : 1399.12 words/s

This was observed on a private Edinburgh branch, but I do not think that the corpus handling code was touched.

**Comment by emjotde** [comment]
Hi, don't think this happened because of right2left. This is applied in the batch level and does not touch actual data. 

Did you maybe run the training processes on the same machine at the same time? Can l see your config?

**Comment by emjotde** [comment]
Maybe we can try to dump the result of the shuffled data somewhere to see what happened.

**Comment by bhaddow** [comment]
Maybe related, but shuffling without sqlite could be swallowing errors. The evidence for this is that I had some runs where the shuffling was happening in /tmp, which caused the root filesystem to fill up making the node (and the cluster admin) unhappy but Marian kept going. With --sqlite Marian crashed when it filled up the filesystem.

**Comment by emjotde** [comment]
Yes, I have similar suspicions.

**Comment by XapaJIaMnu** [comment]
I experienced the same issue (always happens with right-left, but only after second epoch on left-right and seems to be dependent on the random seed). I have a few systems on csd3 on which i can reliably reproduce the issue with left-right.

I would imagine that if disk space was issue, we would experience it every time, not dependent on the random seed, mm?

**Comment by emjotde** [comment]
Oh? OK. I will stare hard at the R2L option. Weird. I have used it all the time for my system.

**Comment by snukky** [comment]
I've got training settings and data from @XapaJIaMnu and I'm checking if I can reproduce the bug.

@bhaddow did you use symbolic links for the training data files in this experiment?

**Comment by bhaddow** [comment]
Symbolic links - Yes I would have used them.

On 28 May 2018 12:46:22 CEST, Roman Grundkiewicz <notifications@github.com> wrote:
>I've got training settings and data from @XapaJIaMnu and I'm checking
>if I can reproduce the bug.
>
>@bhaddow did you use symbolic links for the training data files in this
>experiment?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/marian-nmt/marian-dev/issues/240#issuecomment-392490865

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
The University of Edinburgh is a charitable body, registered in
Scotland, with registration number SC005336.

**Comment by emjotde** [comment]
I saw similar behavior when my temp space filled up. I think this is mostly a matter of missing error messages and bad error handling.

**Comment by snukky** [comment]
I wasn't able to reproduce the bug with the data I've got. Using local storage for temp directory, if you ran out of the disk space before starting/during shuffling, the error message is pretty clear.

**Comment by emjotde** [comment]
I think there is a reason it happened for people during later epochs. If I am not wrong I saw the following for large training data:

* It fit on tempspace during the first shuffle
* During the second shuffle the first shuffled data is still around, space is not freed,
* Second shuffle gets messed up. 

Not sure how reproducible this is.

**Comment by emjotde** [comment]
@snukky 

bumping this up. I saw this situation recently when training two systems on one machine in parallel. One nearly filled up /tmp, the other process barely had any space left but went through without an error message. This resulted in a cut-off epoch (20M instead of 30M).

Working hypothesis: we are not throwing exceptions on IO errors for streams:
http://www.cplusplus.com/reference/ios/ios/exceptions/

**Comment by emjotde** [comment]
@snukky Where are we on this?

**Comment by emjotde** [comment]
We have implemented various checks for eof and bad bits etc. Let's hope we caught this. Re-open if you run into that.
